lib/ecto/query.ex

```diff
@@ -2667,6 +2667,30 @@ defmodule Ecto.Query do
                  where: l.inserted_at > c.updated_at,
                  preload: [:author, comments: {c, likes: l}]
 
+  ## Choosing Between Preloading with Joins vs. Separate Queries
+
+  Deciding between preloading associations via joins, a single large
+  query, (`preload: [comments: c]`) or separate smaller queries
+  (`preload: [:comments]`) depends on the specific use case.
+  Here are some factors to guide your decision:
+
+  * **Joins reduce database round trips:** By fetching data in a single
+  query, joins can minimize database round trips, potentially reducing
+  overall latency.
+  * **Potential for data duplication:** Joins may lead to duplicated
+  data in the result set, which requires more processing by Ecto
+  and consumes more bandwidth when transmitting the results.
+  * **Parallelism with separate queries:** When using separate queries
+  outside of a transaction, Ecto can parallelize the preload queries,
+  which can speed up the overall operation.
+
+  In general, a good default is to only use joins in preloads if you're
+  already joining the associations in the main query. For example,
+  in the last query in the section above, comments and likes are already
+  joined, so they are included in the preload.
+  However, the author is not joined in the main query, so it is preloaded
+  via a separate query.
```

Comment:
Sorry one more comment =(. Please correct me if I'm misunderstanding, but this kind of reads backwards to me. 

i.e. we are already not joining so just use a separate query preload. But it doesn't necessarily answer whether it would have been better to write the main query as a join with authors to begin with.

If it is a 1:1 relationship between posts and authors I feel like it would be more performant to do the join. In this case there is no duplication of the main rows. So you avoid the separate query and in a lot of cases will be doing the combining of the data on the more performant database server.

Reply:
This is again I think quite a nuanced topic. In general I think the advice to use joins for 1:1 and separate query for 1:many is a good starting point. But different cases may yield different results.

For instance if you are joining an enormous amount of of tables it may actually be faster in some cases to take advantage of the parallel preloader. But then again some databases, like postgres, are able to spawn workers to do the query in parallel. But then you have to be careful about memory issues if you use too much parallelism on the database side.

I would personally give the advice to start with join = 1:1, separate query = 1:many and then benchmark. I'm not sure we can do any better than that. 

But again it would be nice to hear other opinions like yours and @josevalim 's

Reply:
@greg-rychlewski you are right and the 1:1 distinction vs. 1:many distinction is a good one.

My main guidance (so far) is: Don't add joins just to preload. If you do that, you should be _sure_ it makes things faster/matters (as it's also more code). 

So, that's why I included it like this in the example. Everything else may also depend too much on the underlying database engine.

Overall, this all may be more a topic for a more detailed book (and it has a 1-2 page discussion in programming ecto) but I still think it's worth giving people some guidance. My main worry is that I've seen queries with 6+ joins added on just for the purpose of preloading and I truly believe that ain't the way :)

Reply:
I might be misunderstanding. When you say adding joins just for preloading what do you mean? If you do a join and want that table's information you won't get it returned without the preload.

Reply:
Are you saying you don't advise doing joins for information retrieval purposes? That it needs to have another purpose like filtering?

Reply:
Yes, that's my overall advice right now for ecto. If I just want to retrieve records and populate their associations a normal `preload` is enough/better imo (due to parallelism and the other problems we discussed).

That said, I'm not an omega DB expert. The docs also adhere by this: every time a join is used for a preload it also fulfills some other function in the query. 

To make it an example:

```
# good
Repo.all from p in Post,
           join: c in assoc(p, :comments),
           where: c.published_at > p.updated_at,
           preload: [comments: c]

Repo.all from p in Post,
           preload: :comments

# bad
Repo.all from p in Post,
           join: c in assoc(p, :comments),
           preload: [comments: c]
```

It's of course more nuanced than that (and in general I'd rather use `Repo.preload` then but different discussion) but that's the overall guideline I'd give people. 

But I also need to rewrite it for clarity if that wasn't clear :)

Reply:
That's my general guideline too: prefer preloads, use joins only if you need them. And Ecto's syntax literally push towards that: loading associations with joins is doing the preload syntax plus a specified join.

Reply:
@greg-rychlewski just general guidelines as the "default place to start" imo. Everything else is where I'd want to see benchmarks or some more arguments. The "Bad" here wasn't meant as "always/usually bad" but as "bad in terms of what I'd recommend as a default" - they can still be very good/useful depending on the circumstances.

@josevalim thanks for chiming in!

Reply:
Understood. I would personally start with 1:1 vs 1:many but just wanted to see if someone like Jose or another maintainer who has been here longer thinks this is ok starting advice. So even though it's not clicking with me personally I at least feel comfortable approving.